src/puny.cpp

```diff
@@ -0,0 +1,385 @@
+    *all++ = c;
+    char_pointer++;
+    continue;
+  }
+  uint32_t number_of_chars(uint32_t(all - all_buffer));
+  uint32_t basic_count = uint32_t(out.size() - 4);
+
+  if (non_basic == non_basic_buffer) { // pure ASCII
+    return out.substr(4);
+  }
+
+  if (basic_count > 0) {
+    out.push_back('-');
+  }
+  uint32_t n = 128;
+  uint32_t bias = 72;
+  uint32_t delta = 0;
+
+  auto sort_unique_values = [](uint32_t array[], size_t size) {
+    size_t duplicates = 0;
+    for (size_t k = 1; k < size; k++) {
+      size_t z = k - duplicates;
+      uint32_t key = array[k];
+      for (; (z >= 1) && (array[z - 1] >= key); z--) {
+      }
+      if (z == k) {
+        // nothing to do!
+      } else if ((array[z] > key)) {
+        std::memmove(array + z + 1, array + z,
+                     (k - duplicates - z) * sizeof(uint32_t));
+        array[z] = key;
+      } else if (array[z] == key) {
+        duplicates++;
+      } else {
+        array[z] = key;
+      }
+    }
+  };
+
+  sort_unique_values(non_basic_buffer, non_basic - non_basic_buffer);
+  non_basic = non_basic_buffer;
+
+  for (uint32_t processed = basic_count; processed < number_of_chars;
+       ++n, ++delta) {
+    uint32_t non_ascii_code_point = *non_basic++;
+    delta += (non_ascii_code_point - n) * (processed + 1);
+    n = non_ascii_code_point;
+    for (size_t i = 0; i < number_of_chars; i++) {
+      uint32_t c = all_buffer[i];
+      if (c < n && (++delta == 0)) { // overflow
+        return std::nullopt;
+      }
+      if (c == n) {
+        for (uint32_t q = delta, k = 36;; k += 36) {
+          uint32_t t = k <= bias ? 1 : (k >= bias + 26 ? 26 : k - bias);
+          if (q < t) {
+            out.push_back(uint8_t(q < 26 ? q + 97 : q + 22));
+            break;
+          }
+          uint32_t char_value = t + (q - t) % (36 - t); // unfortunate division
+          out.push_back(uint8_t(char_value < 26 ? char_value + 97 : char_value + 22));
+
+          q = (q - t) / (36 - t);
+        }
+
+        bias = adapt(delta, processed + 1, basic_count == processed);
+        delta = 0;
+        processed++;
+      }
+    }
+  }
+
+  return out;
+}
+
+// This function merely verifies that a punycode is valid.
+// Important: we do not bother decoding to a buffer.
+bool verify_punycode(std::string_view input) {
+  // So we ignore the first 'xn--'
+  input.remove_prefix(4);
+  if (input.empty()) {
+    return false;
+  }
+  auto loc = input.rfind('-');
+  uint32_t count = 0;
+  if (loc != std::string_view::npos) {
+    for(uint8_t c : input.substr(0, loc)) {
+      if ((c >= 128) || ada::unicode::is_forbidden_domain_code_point(c)) {
+        return false;
+      }
+    }
+    count = uint32_t(loc);
+    input.remove_prefix(loc + 1);
+  }
+  uint32_t n = 128;
+  int i = 0;
+  int bias = 72;
+  for (auto iterator = input.begin(); iterator != input.end();) {
+    int start_i = i;
+    int w = 1;
+    for (int k = 36;; k += 36) {
```

Comment:
We can use simd for this. Nice 👍

Reply:
The punycode algorithm is not easily vectorizable. It is designed from the ground up for character-by-character processing. I am not saying it could not be accelerated, but it would be a major challenge.